Add support for residual predicates in hash joins

EslamAhmed171 opened the pull request:
## Summary

This PR adds support for **mixed predicates** in hash joins, allowing any query with at least one equality condition to be planned as a HASH JOIN (ASOF joins excluded). Previously, joins with **complex expressions** in the ON clause (beyond simple column comparisons) were forced to use slower join algorithms.

**Before this PR:**
```sql
SELECT * FROM t1 LEFT JOIN t2 ON t1.a = t2.b AND t1.x + t2.y > 100
-- Uses Blockwise NL join
```

**After this PR:**
```sql
SELECT * FROM t1 LEFT JOIN t2 ON t1.a = t2.b AND t1.x + t2.y > 100
-- Uses hash join with residual predicate (t1.x + t2.y > 100)
```

### Example Query Plans

**Before:**
```
BLOCKWISE_NL_JOIN
├── SEQ_SCAN t1
└── SEQ_SCAN t2
Conditions: (t1.a = t2.b AND t1.x + t2.y > 100)
```

**After:**
```
HASH_JOIN
├── SEQ_SCAN t1
└── SEQ_SCAN t2
Conditions: t1.a = t2.b
Residual Predicate: t1.x + t2.y > 100
```

## Changes

### 1. Hash Join Support for Residual Predicates

- Modified join creation logic to split ON clause conditions into:
  - **Join conditions** (=, <, >, <=, >=, etc.): used for hash table build/probe and evaluated first
  - **Residual predicate** (everything else): expressions that cannot be pushed down as join conditions, such as:
    - Single-table filters in LEFT JOIN (e.g., `l.db = "DuckDB"`)
    - Complex expressions involving both sides (e.g., `t1.x + t2.y > 100`)

### 2. Join Reorder Optimizer Updates (INNER joins only)

Updated the join reorder optimizer to allow reordering joins with residual predicates for **INNER joins**. 

**Note:** Join reordering with residual predicates is **disabled for SEMI and ANTI joins for NOW**.

### 3. Filter Pushdown for LEFT JOIN

Added check for unsatisfiable residual predicates in LEFT JOIN to convert them to dummy scans (this was failing before: `test/sql/join/left_outer/left_join_issue_15316.test`). This keeps the same behavior as the current code for edge cases.

### 4. Compressed Materialization Optimizer

The compressed materialization optimizer is **disabled for residual predicates for NOW**.

## Future Work

- **Push Filter to Hash Join rule**: We can now add a rule to push filters into hash joins (similar to push filter to table scan). This would be a follow-up optimization.

## Related Issues

Fixes:
- https://github.com/duckdb/duckdb/issues/13953
- https://github.com/duckdb/duckdb/issues/13718
- https://github.com/duckdb/duckdb/issues/10037

EslamAhmed171 committed:
Extract arbitrary expressions as residual predicate in ON clause in CASE 3 to support HASH JOIN later

EslamAhmed171 committed:
Add column binding resolution for comparison join predicates

EslamAhmed171 committed:
Disable column compression for joins with residual predicates

EslamAhmed171 committed:
Disable join reorder optimizer for joins with residual predicates

EslamAhmed171 committed:
Include residual predicates in filter pushdown for inner joins

EslamAhmed171 committed:
Implement residual predicate evaluation in hash join execution

External hash join with residual predicates may be broken

Add support for evaluating mixed predicates during hash join probing.
This enables hash joins for queries with mixed equality and non-equality
conditions in the ON clause (e.g., ON a.x = b.y AND a.z + b.w > 100).

Changes:
- Copy residual predicate from logical to physical operator
- Track probe/build column indices and mappings for predicate evaluation
- Add ApplyResidualPredicate to filter hash matches after equality check
- Integrate predicate evaluation in all join types (inner/outer/mark/single)
- Support residual predicates in external hash join with spilling (needs rechecking)

EslamAhmed171 committed:
WIIIIIIIIIIP

EslamAhmed171 committed:
Cleanup: remove redundant mapping parameters and use stored mappings directly

EslamAhmed171 committed:
WIIIIIIIIP

EslamAhmed171 committed:
Merge branch 'main' into improve-hash-join

EslamAhmed171 committed:
Add comprehensive tests for hash join with residual predicates covering all join types

EslamAhmed171 committed:
NIT

EslamAhmed171 committed:
Allow mixed predicates in join reorder optimizer for INNER joins

EslamAhmed171 committed:
It was not a NIT :( and minor formatting

lnkuiper commented on src/execution/operator/join/perfect_hash_join_executor.cpp:
```diff
@@ -64,6 +64,10 @@ bool ExtractNumericValue(Value val, hugeint_t &result) {
 
 bool PerfectHashJoinExecutor::CanDoPerfectHashJoin(const PhysicalHashJoin &op, const Value &min, const Value &max) {
+	if (op.residual_predicate) {
+		return false;
```
Can you add a `TODO` comment here? Should be doable in the future

lnkuiper commented on src/execution/operator/join/physical_hash_join.cpp:
```diff
@@ -68,45 +87,126 @@ PhysicalHashJoin::PhysicalHashJoin(PhysicalPlan &physical_plan, LogicalOperator
 		rhs_output_columns.col_types.push_back(rhs_col_type);
 	}
+
+	build_input_to_layout_map = std::move(build_input_to_layout);
```
The constructor of `PhysicalHashJoin` is now >150 lines. Would it be possible to move the code related to the residual predicate into a separate initialization function?

lnkuiper commented on src/execution/join_hashtable.cpp:
```diff
@@ -910,16 +926,125 @@ idx_t ScanStructure::ResolvePredicates(DataChunk &keys, SelectionVector &match_s
+	// execute the residual predicate
+	DataChunk result_chunk;
+	result_chunk.Initialize(ht.context, {LogicalType::BOOLEAN});
+	residual_executor->Execute(eval_chunk, result_chunk);
```
Instead of initializing these chunks here each time on the fly, could we store them in `JoinHashTable::ProbeState` instead? This allows us to just initialize them once, and `DataChunk::Reset()` them each time we use them. If you put them as `unique_ptr<ResidualPredicateProbeState>`, you can optionally initialize the chunks. This should be more efficient and clean up this function

lnkuiper reviewed (changes requested):
Thanks for the PR! We have been wanting to do this for a while, but hadn't found the time. This looks great, and should speed up specific queries by orders of magnitude! :)

This is quite a bit diff, so I think this should be looked at by more people than just me. Either way, I think I can give the first round of feedback. I've left a bunch of comments below. Most of them are about cleaning up the code - the functionality looks great already.

I also have a question about your Future Work (I think it's already implemented!):

> Push Filter to Hash Join rule: We can now add a rule to push filters into hash joins (similar to push filter to table scan). This would be a follow-up optimization.

Isn't this already done? You've modified `LogicalComparisonJoin::CreateJoin` to do this properly, and when following the code path in `FilterPushdown`, we actually trigger this code.
```
FilterPushdown::PushdownInnerJoin -> FilterPushdown::PushdownCrossProduct -> LogicalComparisonJoin::CreateJoin
```
Could you add a test that tests this behaviour? Then we can see if it works. If it does not yet work, could you add a `TODO` in the appropriate place in the code for this?

EslamAhmed171 committed:
Add TODO

EslamAhmed171 committed:
Add ResidualPredicateInfo struct to PhysicalHashJoin

EslamAhmed171 committed:
Add extraction utility to PhysicalHashJoin

EslamAhmed171 committed:
Change constructor signature

EslamAhmed171 committed:
Migrate all residual fields to ResidualPredicateInfo

EslamAhmed171 committed:
Refactor PhysicalHashJoin constructor

EslamAhmed171 committed:
Add ResidualPredicateInfo to JoinHashTable

EslamAhmed171 committed:
Add ResidualPredicateProbeState to ScanStructure

EslamAhmed171 committed:
Initialize ResidualPredicateProbeState in ScanStructure

EslamAhmed171 committed:
Improve ApplyResidualPredicate with chunk reuse

EslamAhmed171 committed:
We should avoid planning range joins in case of predicate

EslamAhmed171 commented:
@lnkuiper Thanks for the review!
I’ve pushed a batch refactoring `PhysicalHashJoin` and `JoinHashTable.`
I’ll push another batch later today.

lnkuiper commented:
> I modified the join reorder optimizer to split predicates as some expressions can be used to create edges.

> We could make the optimizer aware of residual predicates and add them back as residual predicates if they cannot be used to create edges.

Nice work, and great idea! Should be somewhere in `QueryGraphManager::GenerateJoins` if I'm not mistaken. This should alleviate the need to do `PushFilter` later.

EslamAhmed171 committed:
Enable column compression

EslamAhmed171 committed:
Unify code paths

EslamAhmed171 committed:
Preserve residual predicates on hash joins in join reorder optimize

EslamAhmed171 commented:
@lnkuiper Good morning! Addressed all your comments, ready for another review round.

EslamAhmed171 committed:
Fix lost residual predicates in join reorder optimizer

Residual predicates that couldn't be attached to joins (e.g., when joins became
cross products) are now pushed as FILTER operators instead of being lost.

lnkuiper reviewed (commented):
Thank you for incorporating the feedback so quickly. It looks excellent.

This is good to go from my side, but I'd like another pair of eyes on this, so I'll request another review, and then we can hopefully get this merged next week :)

EslamAhmed171 committed:
NITs

EslamAhmed171 commented:
> This is good to go from my side, but I'd like another pair of eyes on this, so I'll request another review, and then we can hopefully get this merged next week :)

Thanks for quick response!
I will remove outdated comments and small nit and will open CI here as it continues to be green on my fork.

Mytherin reviewed (commented):
Thanks for the PR! Cool feature. I've left some comments.

This could use some more testing as well. In particular in my experience column ids can be gnarly - so it would be good to have tests that involve expressions that e.g.:

* Use multiple columns from the LHS
* Use multiple columns from the RHS
* Use the same column multiple times from the LHS / RHS
* Use a subset of columns that are in random positions in a larger table (e.g. use columns 6, 9, 12 from `lineitem`)

Adding some of the tests from the linked issues would be great as well.

I also wonder if this always improves performance - e.g. if we have an equality predicate that has many duplicates, could this (significantly) degrade performance compared to a blockwise NL join? Running some benchmarks would be very interesting as well.

Mytherin committed:
Fix #20410: fix for RIGHT SEMI/ANTI - cannot fully label chain as found if there are non-equality predicates present in the join condition

Mytherin committed:
Re-add optimization

EslamAhmed171 commented on test/sql/join/hash_join/hash_join_residual_predicates.test:
```diff
@@ -0,0 +1,327 @@
+PRAGMA enable_verification
+
+statement ok
+CREATE TABLE orders (id INT, customer_id INT, amount INT, status VARCHAR);
```
I created orders and lineitem tables using the TPC-H schema and populated them with custom test data. I did not use dbgen(sf=0.1) to avoid loading default TPC-H data.

Would you prefer I use dbgen(sf=0) (empty tables) followed by custom inserts instead?

EslamAhmed171 replied:
Update:
Used dbgen(sf=0)

EslamAhmed171 committed:
Store arbitrary join expressions in JoinCondition instead of separate predicate field

- Extended JoinCondition to support non-comparison expressions (IsComparison() returns false)
- Removed predicate field from LogicalComparisonJoin, now stored as non-comparison JoinConditions
- Updated all join operators (hash join, nested loop join, ASOF join) to handle new API
- Updated optimizers and column binding resolver to check IsComparison() before accessing left/right
- Added TPCH schema tests for hash join residual predicates

EslamAhmed171 committed:
Resolve conflict

EslamAhmed171 committed:
Make tidy happy

EslamAhmed171 committed:
Cleaning

EslamAhmed171 committed:
Make tidy happy again

EslamAhmed171 commented on src/execution/physical_plan/plan_comparison_join.cpp:
```diff
@@ -50,14 +50,20 @@ PhysicalOperator &PhysicalPlanGenerator::PlanComparisonJoin(LogicalComparisonJoi
 	}
+
+	// for now, only Hash join supports residual predicates (e.g., predicate=l.val+r.val>100)
+	if (op.predicate) {
```
After the fix of `predicate` [here](https://github.com/duckdb/duckdb/pull/20456). I safely removed it

EslamAhmed171 commented:
Hi @Mytherin :)
I've made the following changes:

- Removed predicate from `LogicalComparisonJoin` and updated the `JoinCondition` API
- Added `predicate` to `PhysicalComparisonJoin` since it is used by multiple physical operators (hash join, nested loop join) and makes it easier for future physical plans to support arbitrary predicates

The PR is now ready for another review.

Commit:
Internal #7121: IEJoin Arbitrary Predicates

* Add support for arbitrary predicate evaluation in outer IEJoins
* Not ready for prime time because all comparison join operators have to support this.

Commit:
Internal #7122: PWMJ Arbitrary Predicates

* Add support for arbitrary predicate evaluation in outer PiecewiseMergeJoins
* Not ready for prime time because all comparison join operators have to support this.

hawkfish commented on src/execution/physical_plan/plan_comparison_join.cpp:
```diff
@@ -50,11 +50,17 @@ PhysicalOperator &PhysicalPlanGenerator::PlanComparisonJoin(LogicalComparisonJoi
 	}
+
+	// for now, only Hash join supports residual predicates (e.g., predicate=l.val+r.val>100)
+	if (has_arbitrary_conditions) {
```
The code for the other joins is not hard - I have it implemented.

hawkfish commented on src/storage/serialization/serialize_logical_operator.cpp:
```diff
@@ -328,7 +328,6 @@ void LogicalComparisonJoin::Serialize(Serializer &serializer) const {
 	serializer.WritePropertyWithDefault<vector<LogicalType>>(205, "mark_types", mark_types);
 	serializer.WritePropertyWithDefault<vector<unique_ptr<Expression>>>(206, "duplicate_eliminated_columns", duplicate_eliminated_columns);
 	serializer.WritePropertyWithDefault<bool>(207, "delim_flipped", delim_flipped, false);
-	serializer.WritePropertyWithDefault<unique_ptr<Expression>>(208, "predicate", predicate);
```
Will this break backwards compatibility of stored views?

hawkfish replied:
Ugh I don't know. @Mytherin ?

EslamAhmed171 replied:
Found it should be like this:
```json
{
  "id": 208,
  "name": "predicate",
  "type": "Expression*",
  "status": "deleted"
}
```
and we would end up deserialize the deleted value:
```cpp
deserializer.ReadDeletedProperty<unique_ptr<Expression>>(208, "predicate");
```

Commit:
Internal #7121: IEJoin Arbitrary Predicates

* Also Internal #7122: PWMJ Arbitrary Predicates
* Remove AnyJoin planning for comparison joins
* Update plan checks to look for the residual condition

EslamAhmed171 commented on src/execution/operator/join/physical_piecewise_merge_join.cpp:
```diff
@@ -645,8 +645,8 @@ OperatorResultType PhysicalPiecewiseMergeJoin::ResolveComplexJoin(ExecutionConte
-					tail_count =
-					    SelectJoinTail(conditions[cmp_idx].comparison, left, right, sel, tail_count, &state.sel);
+					tail_count = SelectJoinTail(conditions[cmp_idx].GetComparisonType(), left, right, sel, tail_count,
+					                            &state.sel);
```
Same as above comment

EslamAhmed171 commented on src/include/duckdb/planner/joinside.hpp:
```diff
@@ -16,7 +16,110 @@ namespace duckdb {
+		if (!IsComparison()) {
+			throw InternalException("Swap used on a JoinCondition that is not a left/right comparison");
+		}
+		std::swap(left, right);
```
Nice catch! It doesn't cause a bug currently because I believe it's only used when the ComparisonType is equality, but it is error-prone.

EslamAhmed171 commented:
> @EslamAhmed171 would you mind if I added the residual evaluation for IEJoin and PiecewiseMergeJoin?

Yes, please go ahead! Feel free to add the residual evaluation for IEJoin and PiecewiseMergeJoin.

hawkfish commented:
On my TODO list is improving the planning rules for inequality joins based on statistics and recent improvements to the operators (sorting and threading). So all this simplification is really awesome!

EslamAhmed171 commented on src/planner/binder/tableref/plan_joinref.cpp:
```diff
@@ -270,90 +286,67 @@ unique_ptr<LogicalOperator> LogicalComparisonJoin::CreateJoin(JoinType type, Joi
 
-	// Case 4: Has join conditions but not arbitrary expressions - use comparison join
+	if (!has_equality && !non_comparison_conditions.empty()) {
+		// No equality and has non-comparison - use any join
```
> Have a look at my PR with the other residuals?

Thanks, I'll take  a look now

EslamAhmed171 committed:
Fix review comments: Swap comparison flip, remove arbitrary conditions check, mark predicate as deleted

EslamAhmed171 committed:
Resolve conflicts

hawkfish reviewed (approved):
As long as the tests pass!

EslamAhmed171 committed:
Conflict

lnkuiper reviewed (approved):
Failing CI is unrelated (and is already fixed on `main`). Thanks for all the changes :)

baolinhuang commented:
When will this optimization be released with the official version?

Mytherin commented:
This will be part of the next feature release which is scheduled for fall - see [the release calendar](https://duckdb.org/release_calendar)
